cmd/topicctl/subcmd/rebalance.go

```diff
@@ -0,0 +1,275 @@
+	}
+
+	return nil
+}
+
+func rebalanceRun(cmd *cobra.Command, args []string) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		<-sigChan
+		cancel()
+	}()
+
+	// Keep a cache of the admin clients with the cluster config path as the key
+	adminClients := map[string]admin.Client{}
+	defer func() {
+		for _, adminClient := range adminClients {
+			adminClient.Close()
+		}
+	}()
+
+	clusterConfigPath := rebalanceConfig.shared.clusterConfig
+	topicConfigDir := rebalanceConfig.pathPrefix
+	clusterConfig, err := config.LoadClusterFile(clusterConfigPath, applyConfig.shared.expandEnv)
+	if err != nil {
+		return err
+	}
+	log.Debugf("clusterConfig: %+v", clusterConfig)
+
+	adminClient, ok := adminClients[clusterConfigPath]
+	if !ok {
+		adminClient, err = clusterConfig.NewAdminClient(
+			ctx,
+			nil,
+			applyConfig.dryRun,
+			applyConfig.shared.saslUsername,
+			applyConfig.shared.saslPassword,
+		)
+		if err != nil {
+			return err
+		}
+		adminClients[clusterConfigPath] = adminClient
+	}
+
+	// get all topic configs from path-prefix i.e topics folder
+	// this folder should only have topic configs
+	// recursive search is not performed in the path-prefix
+	log.Infof("Getting all topic configs from path prefix (from root) %v", topicConfigDir)
+	matches, err := filepath.Glob(topicConfigDir + "/**")
+	if err != nil {
+		return err
+	}
+
+	// iterate through each topic config and initiate rebalance
+	topicConfigs := []config.TopicConfig{}
+	for _, match := range matches {
+		topicConfigs, err = config.LoadTopicsFile(match)
+		if err != nil {
+			return err
+		}
+		for _, topicConfig := range topicConfigs {
+			log.Debugf("topicConfig from file: %+v", topicConfig)
+			log.Infof(
+				"Rebalancing topic %s in config %s with cluster config %s",
+				topicConfig.Meta.Name,
+				match,
+				clusterConfigPath,
+			)
+			if err := rebalanceApplyTopic(ctx, topicConfig, clusterConfig, adminClient); err != nil {
+				log.Errorf("Ignoring topic %v for rebalance. Got error: %+v", topicConfig.Meta.Name, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// Perform rebalance on a topic. returns error if unsuccessful
+// topic will not be rebalanced if
+//   - topic config is inconsistent with cluster config (name, region, environment etc...)
+//   - partitions of a topic in kafka cluster does not match with topic partition setting in topic config
+//   - retention.ms of a topic in kafka cluster does not match with topic retentionMinutes setting in topic config
+//
+// to ensure there are no disruptions to kafka cluster
+// topics that are not present in kafka cluster will be applied
+func rebalanceApplyTopic(
+	ctx context.Context,
+	topicConfig config.TopicConfig,
+	clusterConfig config.ClusterConfig,
+	adminClient admin.Client,
+) error {
+	// validate topic config is valid for the cluster config
```

Comment:
Can you elaborate on this please. I didnt understand :(

Reply:
> What about taking the common functionality and putting them into helper functions that are used in each function? That way one have one blessed way of doing each check

Made some changes. Review and resolve please

Reply:
TBH, its not duplicate code.

Even if I could break checkTopic, results is a dict that maintains all the output for each check. Which is something we don't need! 

Or am I missing something? 

three steps in [here](https://github.com/segmentio/topicctl/pull/137/commits/e32bfd2fa72988dc356450037d8c7f3a0b892c8f#diff-0426c16f1a7e7a57632a7ad1158493bb5ecb0a35275ce26dacb537d549271d47R224) are unit already

I still feel this is simple. Also, if we feel like adding more checks in future, Then we can consider breaking the checkTopic functionality

I am trying to avoid checkTopic unless it is needed.